doc/calcelestial.1
.TH CALCELESTIAL "1" "May 2013" "calcelestial 0.1" "User Commands"
.SH NAME
calcelestial - calculates positions, rise, set and transit times of celestial bodies
.SH DESCRIPTION
Together with tools like 'at', 'cron' and 'date' it can be used to schedule arbitrary tasks at planet and moon rise, set or transit times.
.SH SYNOPSIS
.B calcelestial -p [object] -q [location] -m [moment] -f [format]
.SH OPTIONS
.TP
.B -p, --object
available objects are:
.IP
.RS
.RS
sun
.br
moon
.br
mars
.br
neptune
.br
jupiter
.br
mercury
.br
uranus
.br
saturn
.br
venus
.br
pluto
.RE
.RE
.TP
.B -H, --horizon
calc rise/set time with twilight: nautic, civil or astronomical
.TP
.B -t, --time
calc at given time: YYYY-MM-DD [HH:MM:SS]
.TP
.B -m, --moment
calc position at moment of: rise, set, transit
.TP
.B -n, --next
use rise, set, transit time of tomorrow
.TP
.B -f, --format
output format: see \fBstrftime\fR(3) and FORMAT section below for more details
.TP
.B -a, --lat
geographical latitude of observer: -90 to 90deg
.TP
.B -o, --lon
geographical longitude of oberserver: -180 to 180deg
.TP
.B -q, --query
query geonames.org for geographical coordinates
.TP
.B -z, --timezone
override system timezone
.TP
.B -u, --universal
use universial time for parsing and formatting
.TP
.B -l, --local
use the the timezone at --query or --lat / --lon
.TP
.B -h, --help
show this help
.TP
.B -v, --version
show version
.PP
.SH FORMAT
calcelestial supports all conversion specifications as documented in \fBstrftime\fR(3).
.br
additionally these special specifiers have been added:
.TP
.B %J
Julian Date
.TP
.B §r
equatorial right ascension in degrees
.TP
.B §d
equatorial declination in degrees
.TP
.B §a
azimut in degrees from north
.TP
.B §h
altitude in degrees
.TP
.B §d
diameter in arcseconds
.TP
.B §e
distance in kilometer
.TP
.B §A
observer latitude in degrees north
.TP
.B §O
observer longitude in degrees east
.TP
.B §s
azimuth direction as letter, 
.TP
.B §§
A literal '§' character
.SH NOTES
.P
A combination of \fB--lat\fR & \fB--lon\fR or \fB--query\fR is required.
.P
The argument \fB-q, --query\fR fetches coordinates from the geonames.org database. Fetched coordinates will be cached locally. So an active internet connection is only required for the first time.
Please be aware of possible privacy issues!
.P
When symlinking the calcelestial binary to 'sun', 'moon' etc., the argument \fB-p, --object\fR is negligible:
.IP
.B sun -m rise -q Aachen
.SH EXAMPLES
.TP
\fBecho "~/bin/enable-lightning" | at $(calcelestial -p sun -m set -q Frankfurt -H civil)\fR
enable lightning at sunset in Frankfurt
.TP
\fBshutdown $(date -d "+10min $(calcelestial -m transit -a 50.55 -o -6.2)" +%H:%M)\fR
shutdown system 10 minutes after solar noon in Berlin
.TP
\fBnvram-wakeup -s $(date -d "-10min $(calcelestial -m rise -q Aachen)" +%s)\fR
start system 10 minutes before sunrise in Aachen
.SH FILES
geonames.org queries will be cached in \fI~/.geonames.cache\fR
.SH AUTHOR
calcelestial is written by Steffen Vogel <post@steffenvogel.de>
.SH BUGS
.TP
\fB%s\fR formatstring has buggy timezone offset in conjunction with daylight savings


ChangeLog
2016-10-27  Steffen Vogel  <post@steffenvogel.de>

	* use Berkeley DB for on-disk cache
	* improved timezone support
	* added support to query timezone at
	   observer location
	* improved buildsystem & fixed Travis-CI

2012-12-24  Steffen Vogel  <post@steffenvogel.de>

	* removed geonames from make target 'check'
	due to problems with chrooted autobuild networks

2012-12-23  Steffen Vogel  <post@steffenvogel.de>

	* added caching for geonames.org lookups:
	geonames_cache_lookup() and
	geonames_cache_store()

2012-12-22  Steffen Vogel  <post@steffenvogel.de>

	* added a first version of a manpage

2012-11-22  Steffen Vogel  <post@steffenvogel.deA

	* first version


src/calcelestial.c
/**
 * Main routine
 *
 * Does parsing of command line options and start calculation
 *
 * @copyright	2012 Steffen Vogel
 * @license	http://www.gnu.org/licenses/gpl.txt GNU Public License
 * @author	Steffen Vogel <post@steffenvogel.de>
 * @link	https://www.noteblok.net/2012/03/14/cron-jobs-fur-sonnenauf-untergang/
 */
/*
 * This file is part of calcelestial
 *
 * calcelestial is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * calcelestial is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with calcelestial. If not, see <http://www.gnu.org/licenses/>.
 */

#define _XOPEN_SOURCE 700
#define _BSD_SOURCE 1 /* for tm_gmtoff field in struct tm */

#define EXIT_CIRCUMPOLAR 2

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <limits.h>
#include <getopt.h>
#include <float.h>
#include <math.h>
#include <libgen.h>
#include <time.h>
#include <libnova/libnova.h>

#include "../config.h"
#include "objects.h"
#include "formatter.h"
#include "geonames.h"

static struct option long_options[] = {
	{"object",	required_argument, 0, 'p'},
	{"horizon",	required_argument, 0, 'H'},
	{"time",	required_argument, 0, 't'},
	{"moment",	required_argument, 0, 'm'},
	{"next",	no_argument,	   0, 'n'},
	{"format",	required_argument, 0, 'f'},
	{"lat",		required_argument, 0, 'a'},
	{"lon",		required_argument, 0, 'o'},
#ifdef WITH_GEONAMES
	{"query",	required_argument, 0, 'q'},
	{"local",	no_argument,	   0, 'l'},
#endif
	{"timezone",	required_argument, 0, 'z'},
	{"universal",	no_argument,	   0, 'u'},
	{"help",	no_argument,	   0, 'h'},
	{"version",	no_argument,	   0, 'v'},
	{0}
};

static const char *long_options_descs[] = {
	"calc for celestial object: sun, moon, mars, neptune,\n\t\t\t jupiter, mercury, uranus, saturn, venus or pluto",
	"calc rise/set time with twilight: nautic, civil or astronomical",
	"calc at given time: YYYY-MM-DD[_HH:MM:SS]",
	"calc position at moment of: rise, set, transit",
	"use rise, set, transit time of tomorrow",
	"output format: see strftime (3) and calcelestial (1) for more details",
	"geographical latitude of observer: -90° to 90°",
	"geographical longitude of oberserver: -180° to 180°",
#ifdef WITH_GEONAMES
	"query coordinates using the geonames.org geolocation service",
	"query local timezone using the geonames.org geolocation service",
#endif
	"override system timezone (TZ environment variable)",
	"use universial time for parsing and formatting",
	"show usage help",
	"show version"
};

void version()
{
	printf("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
	printf("libnova %s\n", LIBNOVA_VERSION);
}

void usage()
{
	printf("Usage:\n  %s [options]\n\n", PACKAGE_NAME);
	printf("Options:\n");

	struct option *op = long_options;
	const char **desc = long_options_descs;
	while (op->name && desc) {
		printf("  -%c, --%s%s%s\n", op->val, op->name, (strlen(op->name) <= 7) ? "\t\t" : "\t", *desc);
		op++;
		desc++;
	}

	printf("\nA combination of --lat & --lon or --query is required.\n");
	printf("Please report bugs to: %s\n", PACKAGE_BUGREPORT);
}

void usage_error(const char *err)
{
	fprintf(stderr, "Error: %s\n\n", err);
	usage();
	exit(-1);
}

int main(int argc, char *argv[])
{
	int ret;
	time_t t;
	double jd;
	struct tm tm;
	const struct object *obj;

	/* Default options */
	double horizon = LN_SOLAR_STANDART_HORIZON; /* 50 Bogenminuten; no twilight, normal sunset/rise */
	int tz = INT_MAX;

	char *obj_str = basename(argv[0]);
	char *format = "time: %Y-%m-%d %H:%M:%S (%Z) az: §a (§s) alt: §h";
	char tzid[32] = "";
	char *query = NULL;

	bool horizon_set = false;
	bool next = false;
	bool local_tz = false;
	
	time(&t);
	localtime_r(&t, &tm);

	enum {
		MOMENT_NOW,
		MOMENT_RISE,
		MOMENT_SET,
		MOMENT_TRANSIT
	} moment = MOMENT_NOW;

	struct ln_lnlat_posn obs = { DBL_MAX, DBL_MAX };
	struct object_details result;

	/* parse command line arguments */
	while (1) {
		char c = getopt_long(argc, argv, "+hvnult:d:f:a:o:q:z:p:m:H:", long_options, NULL);

		/* detect the end of the options. */
		if (c == -1)
			break;

		switch (c) {
			case 'H':
				if      (strcmp(optarg, "civil") == 0)
					horizon = LN_SOLAR_CIVIL_HORIZON;
				else if (strcmp(optarg, "nautic") == 0)
					horizon = LN_SOLAR_NAUTIC_HORIZON;
				else if (strcmp(optarg, "astronomical") == 0)
					horizon = LN_SOLAR_ASTRONOMICAL_HORIZON;
				else {
					char *endptr;
					horizon = strtod(optarg, &endptr);

					if (endptr == optarg)
						usage_error("invalid horizon / twilight parameter");
				}
				
				horizon_set = true;
				break;

			case 't':
				tm.tm_isdst = -1; /* update dst */
				if (strchr(optarg, '_')) {
					if (!strptime(optarg, "%Y-%m-%d_%H:%M:%S", &tm))
						usage_error("invalid time/date parameter");
				}
				else {
					if (!strptime(optarg, "%Y-%m-%d", &tm))
						usage_error("invalid time/date parameter");
				}
				break;

			case 'm':
				if      (strcmp(optarg, "rise") == 0)
					moment = MOMENT_RISE;
				else if (strcmp(optarg, "set") == 0)
					moment = MOMENT_SET;
				else if (strcmp(optarg, "transit") == 0)
					moment = MOMENT_TRANSIT;
				else
					usage_error("invalid moment");
				break;

			case 'n':
				next = true;
				break;

			case 'f':
				format = strdup(optarg);
				break;

			case 'a':
				obs.lat = strtod(optarg, NULL);
				break;

			case 'o':
				obs.lng = strtod(optarg, NULL);
				break;
#ifdef WITH_GEONAMES
			case 'q':
				query = strdup(optarg);
				break;

			case 'l':
				local_tz = true;
				break;
#endif
			case 'p':
				obj_str = optarg;
				break;

			case 'z':
				strncpy(tzid, optarg, sizeof(tzid));
				break;

			case 'u':
				strncpy(tzid, "UTC", sizeof(tzid));
				break;

			case 'v':
				version();
				return 0;

			case 'h':
				usage();
				return 0;

			case '?':
			default:
				usage_error("unrecognized option");
		}
	}
	
	/* Parse planet/obj */
	obj = object_lookup(obj_str);
	if (!obj)
		usage_error("invalid or missing object, use --object");

#ifdef WITH_GEONAMES
	/* Lookup place at http://geonames.org */
	if (query) {
		ret =  geonames_lookup_latlng(query, &obs, NULL, 0);
		if (ret)
			usage_error("failed to lookup location");
	}
	
	if (local_tz) {
		int gmt_offset;
		ret =  geonames_lookup_tz(obs, &gmt_offset, tzid, sizeof(tzid));
		if (ret)
			usage_error("failed to lookup location");
	}
#endif

	setenv("TZ", tzid, 1);
	tzset();

	/* Validate observer coordinates */
	if (fabs(obs.lat) > 90)
		usage_error("invalid latitude, use --lat");
	if (fabs(obs.lng) > 180)
		usage_error("invalid longitude, use --lon");
	
	if (horizon_set && strcmp(object_name(obj), "sun"))
		usage_error("the twilight parameter can only be used for the sun");

	/* Calculate julian date */
	t = mktime(&tm);
	jd = ln_get_julian_from_timet(&t);

	result.obs = obs;

#ifdef DEBUG
	printf("Debug: calculate for jd: %f\n", jd);
	printf("Debug: calculate for ts: %ld\n", t);
	printf("Debug: for position: N %f, E %f\n", obs.lat, obs.lng);
	printf("Debug: for object: %s\n", object_name(obj));
	printf("Debug: with horizon: %f\n", horizon);
	printf("Debug: with timezone: %s\n", tzid);
#endif

	/* calc rst date */
rst:	if (object_rst(obj, jd - .5, horizon, &result.obs, &result.rst) == 1)  {
		if (moment != MOMENT_NOW) {
			fprintf(stderr, "object is circumpolar\n");
			return 2;
		}
	}
	else {
		switch (moment) {
			case MOMENT_NOW:	result.jd = jd; break;
			case MOMENT_RISE:	result.jd = result.rst.rise; break;
			case MOMENT_SET:	result.jd = result.rst.set; break;
			case MOMENT_TRANSIT:	result.jd = result.rst.transit; break;
		}

		if (next && result.jd < jd) {
			jd++;
			next = false;
			goto rst;
		}
	}
	
	ln_get_timet_from_julian(result.jd, &t);
	localtime_r(&t, &result.tm);

	object_pos(obj, jd, &result);

	format_result(format, &result);

	return 0;
}
